File: st/token_bucket.py
import asyncio
import time


class TokenBucket:
    """
    A thread-safe token bucket rate limiter using asyncio.Lock.
    
    Guarantees:
    - No race conditions: token deduction is atomic within the lock
    - Strict serialization: File 1 cannot read state until File 2 completes deduction
    - Mathematically sound: deficit calculation and wait time are precise
    - Global pause support: 429 errors trigger a 35-second global pause
    """

    def __init__(self, max_rpm=30, max_tpm=30000):
        """
        Initialize the token bucket.
        
        Args:
            max_rpm: Maximum requests per minute (for compatibility)
            max_tpm: Maximum tokens per minute (capacity and refill rate)
        """
        self.max_rpm = max_rpm
        self.max_tpm = max_tpm
        self.capacity = max_tpm
        self.tokens = max_tpm
        self.refill_rate = max_tpm / 60.0  # Tokens per second
        self.last_refill = time.time()
        self.lock = asyncio.Lock()
        self.global_pause_until = 0.0
        
        # Dynamic max_file_tokens based on TPM
        # For 500k TPM: allow up to 100k tokens per file (20% of capacity)
        # For 30k TPM: allow up to 22k tokens per file (73% of capacity)
        # Formula: min(TPM * 0.2, 100000) to cap at 100k for very high TPM
        self.max_file_tokens = min(int(max_tpm * 0.2), 100000)
        
        self.safety_buffer = 1.10  # 10% safety margin

    async def consume(self, requested_tokens):
        """
        Consume tokens from the bucket with strict lock-based serialization.
        
        This method ensures:
        1. Token state is read and modified atomically within the lock
        2. Wait time is calculated before releasing the lock
        3. Sleep happens outside the lock to avoid blocking other tasks
        4. Tokens never go negative and last_refill never advances into the future
        
        Args:
            requested_tokens: Number of tokens to consume
            
        Returns:
            Total wait time in seconds (pause_time + wait_time)
        """
        wait_time = 0.0
        pause_time = 0.0

        # CRITICAL: All state mutations happen inside the lock
        async with self.lock:
            now = time.time()

            # Step 1: Check if we're in a global pause
            if now < self.global_pause_until:
                pause_time = self.global_pause_until - now
                # Don't refill during pause, just return the pause time
                return pause_time

            # Step 2: Refill tokens based on elapsed time
            elapsed = now - self.last_refill
            # Ensure elapsed is never negative (defensive programming)
            if elapsed < 0:
                elapsed = 0
                self.last_refill = now
            
            self.tokens = min(
                self.capacity,
                self.tokens + (elapsed * self.refill_rate)
            )

            # Step 3: Check if we have enough tokens
            if self.tokens >= requested_tokens:
                # We have enough tokens, deduct and return immediately
                self.tokens -= requested_tokens
                self.last_refill = now
                return 0.0

            # Step 4: Calculate deficit and wait time
            deficit = requested_tokens - self.tokens
            wait_time = deficit / self.refill_rate

            # Step 5: FIX - Set tokens to 0 and keep last_refill at NOW
            # This prevents negative tokens and future time advancement
            self.tokens = 0
            self.last_refill = now

        # Step 6: Sleep OUTSIDE the lock (critical for concurrency)
        if wait_time > 0:
            await asyncio.sleep(wait_time)

        return wait_time

    async def trigger_global_pause(self):
        """
        Trigger a global pause when a 429 error is encountered.
        
        This method:
        1. Acquires the lock to ensure atomic state update
        2. Sets global_pause_until to 35 seconds in the future
        3. All subsequent consume() calls will wait until this time passes
        """
        async with self.lock:
            self.global_pause_until = time.time() + 35.0

    async def set_global_pause(self, duration: float = 30.0):
        """
        Alias for trigger_global_pause for backward compatibility.
        Sets global pause flag when 429 error occurs.
        """
        async with self.lock:
            self.global_pause_until = time.time() + duration

    async def consume_with_wait(self, file_tokens: int) -> float:
        """
        Consume tokens, waiting if necessary. Returns actual wait time.
        
        This is the main method called by audit_file.
        Handles the full consume-and-wait flow.
        """
        # Check if file is too large
        if file_tokens > self.max_file_tokens:
            return -1.0  # Signal to skip this file
        
        # Use the main consume method which handles everything
        wait_time = await self.consume(file_tokens)
        return wait_time

    def refund(self, tokens_estimated: int, tokens_actual: int):
        """
        Refund tokens if estimate was too high.
        This is called when actual token usage is less than estimated.
        """
        if tokens_estimated > tokens_actual:
            refund_amount = tokens_estimated - tokens_actual
            self.tokens = min(self.capacity, self.tokens + refund_amount)

    async def get_state(self):
        """
        Get the current state of the token bucket (for debugging/monitoring).
        
        Returns:
            dict with current tokens, capacity, and pause status
        """
        async with self.lock:
            now = time.time()
            elapsed = now - self.last_refill
            current_tokens = min(
                self.capacity,
                self.tokens + (elapsed * self.refill_rate)
            )
            is_paused = now < self.global_pause_until
            return {
                "tokens": current_tokens,
                "capacity": self.capacity,
                "refill_rate": self.refill_rate,
                "is_paused": is_paused,
                "pause_until": self.global_pause_until if is_paused else None,
            }

File: extension/token_bucket.js
/**
 * Token Bucket Rate Limiter (JavaScript Port)
 * 
 * Ported from st/token_bucket.py for client-side rate limiting.
 * Implements the same mathematical deficit calculation as stress_test.py.
 * 
 * Key Features:
 * - Continuous token refill at refill_rate = max_tpm / 60.0 tokens/second
 * - Mathematical deficit calculation: wait_time = deficit / refill_rate
 * - Token refunds for overestimates
 * - Global pause support for 429 errors
 */

class TokenBucket {
    /**
     * Initialize the token bucket.
     * 
     * @param {number} maxRpm - Maximum requests per minute
     * @param {number} maxTpm - Maximum tokens per minute (capacity and refill rate)
     */
    constructor(maxRpm = 30, maxTpm = 30000) {
        this.maxRpm = maxRpm;
        this.maxTpm = maxTpm;
        this.capacity = maxTpm;
        this.tokens = maxTpm;
        this.refillRate = maxTpm / 60.0;  // Tokens per second
        this.lastRefill = Date.now() / 1000;  // Convert to seconds
        this.globalPauseUntil = 0.0;
        
        // Dynamic max_file_tokens based on TPM
        // For 500k TPM: allow up to 100k tokens per file (20% of capacity)
        // For 30k TPM: allow up to 6k tokens per file (20% of capacity)
        this.maxFileTokens = Math.min(Math.floor(maxTpm * 0.2), 100000);
        
        this.safetyBuffer = 1.10;  // 10% safety margin
        
        console.log(`[TOKEN BUCKET] Initialized: ${maxTpm.toLocaleString()} TPM, ${maxRpm} RPM`);
        console.log(`[TOKEN BUCKET] Refill rate: ${this.refillRate.toFixed(2)} tokens/second`);
        console.log(`[TOKEN BUCKET] Max file size: ${this.maxFileTokens.toLocaleString()} tokens`);
    }

    /**
     * Consume tokens from the bucket with mathematical deficit calculation.
     * 
     * @param {number} requestedTokens - Number of tokens to consume
     * @returns {Promise<number>} - Total wait time in seconds
     */
    async consume(requestedTokens) {
        const now = Date.now() / 1000;  // Convert to seconds
        
        // Step 1: Check if we're in a global pause
        if (now < this.globalPauseUntil) {
            const pauseTime = this.globalPauseUntil - now;
            console.log(`[TOKEN BUCKET] Global pause active, waiting ${pauseTime.toFixed(2)}s`);
            await this.sleep(pauseTime * 1000);  // Convert to milliseconds
            return pauseTime;
        }

        // Step 2: Refill tokens based on elapsed time
        const elapsed = now - this.lastRefill;
        this.tokens = Math.min(
            this.capacity,
            this.tokens + (elapsed * this.refillRate)
        );

        // Step 3: Check if we have enough tokens
        if (this.tokens >= requestedTokens) {
            // We have enough tokens, deduct and return immediately
            this.tokens -= requestedTokens;
            this.lastRefill = now;
            return 0.0;
        }

        // Step 4: Calculate deficit and wait time
        const deficit = requestedTokens - this.tokens;
        const waitTime = deficit / this.refillRate;

        console.log(`[TOKEN BUCKET] Deficit: ${deficit.toFixed(0)} tokens, waiting ${waitTime.toFixed(2)}s`);

        // Step 5: Set tokens to 0 and keep last_refill at NOW
        this.tokens = 0;
        this.lastRefill = now;

        // Step 6: Sleep for the calculated wait time
        await this.sleep(waitTime * 1000);  // Convert to milliseconds

        return waitTime;
    }

    /**
     * Consume tokens with wait, checking file size limits.
     * 
     * @param {number} fileTokens - Number of tokens to consume
     * @returns {Promise<number>} - Wait time in seconds, or -1 if file too large
     */
    async consumeWithWait(fileTokens) {
        // Check if file is too large
        if (fileTokens > this.maxFileTokens) {
            return -1.0;  // Signal to skip this file
        }
        
        return await this.consume(fileTokens);
    }

    /**
     * Refund tokens if estimate was too high.
     * 
     * @param {number} tokensEstimated - Estimated tokens
     * @param {number} tokensActual - Actual tokens used
     */
    refund(tokensEstimated, tokensActual) {
        if (tokensEstimated > tokensActual) {
            const refundAmount = tokensEstimated - tokensActual;
            this.tokens = Math.min(this.capacity, this.tokens + refundAmount);
            console.log(`[TOKEN BUCKET] Refunded ${refundAmount} tokens`);
        }
    }

    /**
     * Set global pause when 429 error occurs.
     * 
     * @param {number} duration - Pause duration in seconds
     */
    async setGlobalPause(duration = 30.0) {
        this.globalPauseUntil = (Date.now() / 1000) + duration;
        console.log(`[TOKEN BUCKET] Global pause set for ${duration}s`);
    }

    /**
     * Get current state of the token bucket.
     * 
     * @returns {Object} - Current state
     */
    getState() {
        const now = Date.now() / 1000;
        const elapsed = now - this.lastRefill;
        const currentTokens = Math.min(
            this.capacity,
            this.tokens + (elapsed * this.refillRate)
        );
        const isPaused = now < this.globalPauseUntil;
        
        return {
            tokens: currentTokens,
            capacity: this.capacity,
            refillRate: this.refillRate,
            isPaused: isPaused,
            pauseUntil: isPaused ? this.globalPauseUntil : null
        };
    }

    /**
     * Sleep helper function.
     * 
     * @param {number} ms - Milliseconds to sleep
     * @returns {Promise<void>}
     */
    sleep(ms) {
        return new Promise(resolve => setTimeout(resolve, ms));
    }
}

module.exports = { TokenBucket };
